rsocket-ipc-core/src/main/java/io/rsocket/ipc/Client.java

```diff
@@ -0,0 +1,233 @@
+import reactor.core.publisher.Flux;
+import reactor.core.publisher.Mono;
+
+@SuppressWarnings("unchecked")
+public final class Client<I, O> {
+
+  private final String service;
+  private final Marshaller<I> marshaller;
+  private final Unmarshaller<O> unmarshaller;
+  private final RSocket rSocket;
+
+  private Client(
+      final String service,
+      final Marshaller marshaller,
+      final Unmarshaller unmarshaller,
+      final RSocket rSocket) {
+    this.service = service;
+    this.marshaller = marshaller;
+    this.unmarshaller = unmarshaller;
+    this.rSocket = rSocket;
+  }
+
+  public interface R {
+    P rsocket(RSocket rSocket);
+  }
+
+  public interface P {
+    <I> U<I> marshall(Marshaller<I> marshaller);
+  }
+
+  public interface U<I> {
+    <O> Client<I, O> unmarshall(Unmarshaller<O> unmarshaller);
+  }
+
+  public Functions.RequestResponse<I, O> requestResponse(String route) {
+    Objects.requireNonNull(marshaller);
+    Objects.requireNonNull(unmarshaller);
+    Objects.requireNonNull(rSocket);
+    return (o, byteBuf) ->
+        doRequestResponse(service, route, rSocket, marshaller, unmarshaller, o, byteBuf);
+  }
+
+  public Functions.RequestChannel<I, O> requestChannel(String route) {
+    Objects.requireNonNull(marshaller);
+    Objects.requireNonNull(unmarshaller);
+    Objects.requireNonNull(rSocket);
+    return (publisher, byteBuf) ->
+        doRequestChannel(service, route, rSocket, marshaller, unmarshaller, publisher, byteBuf);
+  }
+
+  public Functions.RequestStream<I, O> requestStream(String route) {
+    Objects.requireNonNull(marshaller);
+    Objects.requireNonNull(unmarshaller);
+    Objects.requireNonNull(rSocket);
+    return (o, byteBuf) ->
+        doRequestStream(service, route, rSocket, marshaller, unmarshaller, o, byteBuf);
+  }
+
+  public Functions.FireAndForget<I> fireAndForget(String route) {
+    Objects.requireNonNull(marshaller);
+    Objects.requireNonNull(unmarshaller);
+    Objects.requireNonNull(rSocket);
+    return (o, byteBuf) -> doFireAndForget(service, route, rSocket, marshaller, o, byteBuf);
+  }
+
+  private static class Builder implements P, U, R {
+    private final String service;
+    private Marshaller marshaller;
+    private Unmarshaller unmarshaller;
+    private RSocket rSocket;
+
+    private Builder(String service) {
+      this.service = service;
+    }
+
+    @Override
+    public <I> U<I> marshall(Marshaller<I> marshaller) {
+      Objects.requireNonNull(marshaller);
+      this.marshaller = marshaller;
+      return this;
+    }
+
+    @Override
+    public Client unmarshall(Unmarshaller unmarshaller) {
+      Objects.requireNonNull(unmarshaller);
+      this.unmarshaller = unmarshaller;
+      return new Client(service, marshaller, unmarshaller, rSocket);
+    }
+
+    @Override
+    public P rsocket(RSocket rSocket) {
+      Objects.requireNonNull(rSocket);
+      this.rSocket = rSocket;
+      return this;
+    }
+  }
+
+  private Mono<Void> doFireAndForget(
```

Comment:
Guessing this should be static. Otherwise redundant parameters passing
```suggestion
  private static <I> Mono<Void> doFireAndForget(
```